src/main/kotlin/org/rust/lang/core/resolve/Processors.kt

```diff
@@ -62,59 +61,102 @@ interface RsResolveProcessorBase<in T : ScopeEntry> {
 
+fun createProcessor(names: Set<String>?, processor: (ScopeEntry) -> Boolean): RsResolveProcessor =
+    createProcessorGeneric(names, processor)
+
 fun <T : ScopeEntry> createProcessorGeneric(
     name: String? = null,
     processor: (T) -> Boolean
+): RsResolveProcessorBase<T> = createProcessorGeneric(name?.let { setOf(it) }, processor)
+
+fun <T : ScopeEntry> createProcessorGeneric(
+    names: Set<String>? = null,
+    processor: (T) -> Boolean
 ): RsResolveProcessorBase<T> {
     return object : RsResolveProcessorBase<T> {
         override fun invoke(entry: T): Boolean = processor(entry)
-        override val name: String? = name
-        override fun toString(): String = "Processor(name=$name)"
+        override val names: Set<String>? = names
+        override fun toString(): String = "Processor(name=$names)"
     }
 }
 
 typealias RsMethodResolveProcessor = RsResolveProcessorBase<MethodResolveVariant>
 
 fun collectPathResolveVariants(
+    ctx: PathResolutionContext,
     path: RsPath,
     f: (RsResolveProcessor) -> Unit
-): List<BoundElementWithVisibility<RsElement>> {
+): List<RsPathResolveResult<RsElement>> {
     val referenceName = path.referenceName ?: return emptyList()
-    val result = SmartList<BoundElementWithVisibility<RsElement>>()
+    val result = SmartList<RsPathResolveResult<RsElement>>()
     val processor = createProcessor(referenceName) { e ->
         if ((e == ScopeEvent.STAR_IMPORTS) && result.isNotEmpty()) {
             return@createProcessor true
         }
 
         if (e.name == referenceName) {
-            val element = e.element ?: return@createProcessor false
-            if (element !is RsDocAndAttributeOwner || element.existsAfterExpansionSelf) {
-                val boundElement = BoundElement(element, e.subst)
-                val visibilityStatus = e.getVisibilityStatusFrom(path)
-                if (visibilityStatus != VisibilityStatus.CfgDisabled) {
-                    val isVisible = visibilityStatus == VisibilityStatus.Visible
-                    result += BoundElementWithVisibility(boundElement, isVisible)
-                }
-            }
+            collectPathScopeEntry(ctx, result, e)
+        }
+        false
+    }
+    f(processor)
+    return result
+}
+
+fun collectMultiplePathResolveVariants(
+    ctx: PathResolutionContext,
+    paths: List<RsPath>,
+    f: (RsResolveProcessor) -> Unit
+): Map<RsPath, SmartList<RsPathResolveResult<RsElement>>> {
+    val result: MutableMap<RsPath, SmartList<RsPathResolveResult<RsElement>>> = hashMapOf()
+    val resultByName: MutableMap<String, SmartList<RsPathResolveResult<RsElement>>> = hashMapOf()
+    for (path in paths) {
+        val name = path.referenceName ?: continue
+        val list = resultByName.getOrPut(name) { SmartList() }
+        result[path] = list
+    }
+    val processor = createProcessor(resultByName.keys) { e ->
+        if ((e == ScopeEvent.STAR_IMPORTS) && result.values.all { it.isNotEmpty() }) {
```

Comment:
Is `STAR_IMPORTS` just an optimization or does it control shadowing?

Reply:
It was an optimization in old name resolution, but now it's nothing but a nonsense. I'll remove it later.
If we'll decide to introduce a similar shadowing control technique, I'd rather add a specific method to `RsResolveProcessorBase`, e.g. `.beginNextScope()`